hail/python/hail/methods/qc.py

```diff
@@ -244,51 +242,47 @@ def variant_qc(mt, name='variant_qc') -> MatrixTable:
 
     if not has_field_of_type('GT',  hl.tcall):
         raise ValueError(f"'variant_qc': expect an entry field 'GT' of type 'call'")
-    exprs['n_called'] = hl.agg.count_where(hl.is_defined(mt['GT']))
-    struct_exprs.append(hl.agg.call_stats(mt.GT, mt.alleles))
-
-
-    # the structure of this function makes it easy to add new nested computations
-    def flatten_struct(*struct_exprs):
-        flat = {}
-        for struct in struct_exprs:
-            for k, v in struct.items():
-                flat[k] = v
-        return hl.struct(
-            **flat,
-            **exprs,
-        )
-
-    mt = mt.annotate_rows(**{name: hl.bind(flatten_struct, *struct_exprs)})
-
-    hwe = hl.hardy_weinberg_test(mt[name].homozygote_count[0],
-                                 mt[name].AC[1] - 2 * mt[name].homozygote_count[1],
-                                 mt[name].homozygote_count[1])
-    hwe = hwe.select(het_freq_hwe=hwe.het_freq_hwe, p_value_hwe=hwe.p_value)
-    mt = mt.annotate_rows(**{name: mt[name].annotate(n_not_called=n_samples - mt[name].n_called,
-                                                     call_rate=mt[name].n_called / n_samples,
-                                                     n_het=mt[name].n_called - hl.sum(mt[name].homozygote_count),
-                                                     n_non_ref=mt[name].n_called - mt[name].homozygote_count[0],
-                                                     **hl.cond(hl.len(mt.alleles) == 2,
-                                                               hwe,
-                                                               hl.null(hwe.dtype)))})
-    return mt
+
+    bound_exprs['n_called'] = hl.agg.count_where(hl.is_defined(mt['GT']))
+    bound_exprs['n_not_called'] = hl.agg.count_where(hl.is_missing(mt['GT']))
+    bound_exprs['n_filtered'] = mt.count_cols(_localize=False) - hl.agg.count()
+    bound_exprs['call_stats'] = hl.agg.call_stats(mt.GT, mt.alleles)
+
+    result = hl.rbind(hl.struct(**bound_exprs),
+                      lambda e1: hl.rbind(
+                          hl.case().when(hl.len(mt.alleles) == 2,
+                                         hl.hardy_weinberg_test(e1.call_stats.homozygote_count[0],
+                                                                e1.call_stats.AC[1] - 2 *
+                                                                e1.call_stats.homozygote_count[1],
+                                                                e1.call_stats.homozygote_count[1])
+                                         ).or_missing(),
+                          lambda hwe: hl.struct(**{
+                              **gq_dp_exprs,
+                              **e1.call_stats,
+                              'call_rate': hl.float(e1.n_called) / (e1.n_called + e1.n_not_called + e1.n_filtered),
+                              'n_called': e1.n_called,
+                              'n_not_called': e1.n_not_called,
```

Comment:
does the order of the fields matter for backwards compatibility?

Reply:
Really any change in schemas could technically break a pipeline, I suppose. I'm comfortable with this, though.